test: Update `test_ensure_mod_tests`

JonathanWoollett-Light opened the pull request:
## Changes

Updates `test_ensure_mod_tests` to allow easier filtering.

## Reason

It allows for easier filtering.

This was spawned from attempting to change the filter in https://github.com/firecracker-microvm/firecracker/pull/3528 and finding it awkward.

This change should also help support further checking:
1. Checking `mod tests` is at the end of the file.
2. Checking files with `mod tests` don't have `#[test]`s outside of this.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following
Developer Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [x] If a specific issue led to this PR, this PR closes the issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] API changes follow the [Runbook for Firecracker API changes][2].
- [x] User-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.
- [x] New `TODO`s link to an issue.
- [x] Commits meet [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

---

- [x] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

pb8o reviewed (commented):
It's not clear what the motivation for the change is. Is it difficult to add new exceptions?

JonathanWoollett-Light commented on tests/integration_tests/style/test_rust.py:
```diff
@@ -21,40 +24,47 @@ def test_rust_style():
+    # Install ripgrep
+    subprocess.run(["cargo", "install", "ripgrep"], capture_output=True, check=True)
+
+    # Files with `mod tests`.
+    mods = cmd('rg -l "mod tests" -g "*.rs"')
+
+    # Files with `#[test]`.
+    tests = cmd('rg -l "#\\[test\\]" -g "*.rs"')
+
+    # Files with `#[test]` without `mod tests`.
+    tests_without_mods = disjoint(tests, mods)
+
```
This is fair but I think the example you have given here is more complex, there are more arguments and tools.

roypat replied:
Is this even what we want? I thought the goal of this test was to ensure that all tests are inside a `mod test` (unless they're integration tests or generated by bindgen.rs`), but with these regex check a `.rs` file with the structure below passes CI, even though it violates the spirit of the test (although the documentation only says "ensure all modules with tests have a `mod test`", which doesn't rule this case as "bad")

```rs
#[test] 
fn bad() {}

#[cfg(test)]
mod tests {}
```

JonathanWoollett-Light replied:
@roypat This issue is present in our current implementation.

roypat replied:
Yep, I know. I just felt that if we're redoing this test anyway, and we agree that this is not the intended behavior, we might as well fix it.

pb8o replied:
If we want to do a more accurate test, I think we should let go of shelling out to either tool and do the check in Python. I.e sthg like (untested)

```python
for file in Path("..").rglob("*.rs):
    match = re.match(r"#\[test\] .*^mod tests {", file.read_text(), re.MULTILINE)
    assert match is None
```

roypat replied:
Thinking about it some more, I think any attempt at solving this using regex breaks on 

```rs
#[cfg(test)]
mod test {}

#[test]
fn bad() {}
```

because a regex cannot know when the `mod test` is over - bracket matching is the sterotypical context free language that is not regular

JonathanWoollett-Light replied:
I think a fully accurate solution would ultimately need to consider the Rust AST (akin to https://docs.rs/syn/latest/syn/struct.File.html). My overall view is this test is not worth the effort, in the worst case a test gets included in coverage and raises coverage slightly higher than it should have, I don't think this is a significant impact.

roypat replied:
yeah, I agree on this.

JonathanWoollett-Light replied:
We came to the conclusion to merge this PR.

JonathanWoollett-Light commented on tests/integration_tests/style/test_rust.py:
```diff
@@ -21,40 +24,47 @@ def test_rust_style():
+    excluding = [
+        "_gen/",
+        "/tests/",
+        "/test_utils",
```
Makes sense to remove it here if it works I agree (took it from https://github.com/firecracker-microvm/firecracker/blob/main/tests/integration_tests/build/test_coverage.py#L78).

JonathanWoollett-Light replied:
Fixed.

JonathanWoollett-Light commented on tests/integration_tests/style/test_rust.py:
```diff
@@ -21,40 +24,47 @@ def test_rust_style():
+    monkeypatch.chdir("..")
+
+    # Install ripgrep
+    subprocess.run(["cargo", "install", "ripgrep"], capture_output=True, check=True)
```
1. This wouldn't significantly impact test performance in an environment where the tool was already installed.
2. In retrospect I agree, I can remove the install (is useful for now to see CI result without needing to update docker container) (I think you meant `but it would *not* be good to add more`)

JonathanWoollett-Light replied:
This no longer applies (as ripgrep has been removed).

JonathanWoollett-Light commented:
> It's not clear what the motivation for the change is. Is it difficult to add new exceptions?

With the previous implementation it did not exclude all directories of `_gen` it required them to be specifically under `/src/` adapting the previous implementation for this became messier.

Although what you suggested in https://github.com/firecracker-microvm/firecracker/pull/3530#discussion_r1136160641 could be used with the Python code suggested here to do this easily.

roypat commented on tests/integration_tests/style/test_rust.py:
```diff
@@ -21,40 +24,47 @@ def test_rust_style():
+    final = [f for f in tests_without_mods if not any(x in f for x in excluding)]
+
+    # Assert `final` is empty.
+    assert (
+        final == []
+    ), "`#[test]`s found in files without `mod tests`s: {final}. Code coverage requires that tests are in test modules."
```
nit: since you do `assert final == []`, python will already print out the contents of `final` if the assertion fails. Printing it again as part of the "Assertion Failed" message will be just noise.

JonathanWoollett-Light commented:
Removed `ripgrep` and updated to use:
```
find ../src -type f -name "*.rs" |xargs grep --files-without-match "mod tests {" |xargs grep --files-with-matches "#\[test\]"
```
from https://github.com/firecracker-microvm/firecracker/pull/3530#discussion_r1136160641

pb8o commented on tests/integration_tests/style/test_rust.py:
```diff
@@ -27,34 +30,28 @@ def test_ensure_mod_tests():
+    ]
+
+    def cmd(string):
+        cmd = shlex.split(string)
```
don't need to `shlex` if you use `shell=True` I think.

JonathanWoollett-Light replied:
Fixed.

pb8o commented on tests/integration_tests/style/test_rust.py:
```diff
@@ -27,34 +30,28 @@ def test_ensure_mod_tests():
+    ]
+
+    def cmd(string):
+        cmd = shlex.split(string)
```
Here the local variable is overriding the function name.

JonathanWoollett-Light replied:
Fixed.

pb8o commented on tests/integration_tests/style/test_rust.py:
```diff
@@ -27,34 +30,28 @@ def test_ensure_mod_tests():
+        "src/io_uring/src/bindings.rs",
+    ]
+
+    def cmd(string):
+        cmd = shlex.split(string)
+        process = subprocess.run(cmd, shell=True, capture_output=True, check=True)
+        return process.stdout.decode("utf-8").split("\n")
+
+    # Files with `#[test]` without `mod tests`.
+    tests_without_mods = cmd(
+        r'find ../src -type f -name "*.rs" |xargs grep --files-without-match "mod tests {" |xargs grep --files-with-matches "#\[test\]"'
     )
```
Don't need the nested function, and I think the `r""` string is messing with shell quoting. Maybe try:

```python
cmd = 'find ../src -type f -name "*.rs" |xargs grep --files-without-match "mod tests {" |xargs grep --files-with-matches "#\[test\]"'
res = subprocess.run(cmd, shell=True, capture_output=True, check=True)
test_without_mods = res.stdout.decode("utf-8").split("\n")
```
